validator/src/main/java/org/ethereum/beacon/validator/proposer/BeaconChainProposerImpl.java

```diff
@@ -0,0 +1,170 @@
+    Bytes8 domain = specHelpers.get_domain(state.getForkData(), state.getSlot(), PROPOSAL);
+    return signer.sign(proposalRoot, domain);
+  }
+
+  private Bytes96 getRandaoReveal(BeaconState state, UInt24 index, MessageSigner<Bytes96> signer) {
+    ValidatorRecord proposer = state.getValidatorRegistry().get(SpecHelpers.safeInt(index));
+    Hash32 hash = Hash32.wrap(Bytes32.leftPad(proposer.getProposerSlots().toBytesBigEndian()));
+    Bytes8 domain = specHelpers.get_domain(state.getForkData(), state.getSlot(), RANDAO);
+    return signer.sign(hash, domain);
+  }
+
+  /*
+   Let D be the set of Eth1DataVote objects vote in state.eth1_data_votes.
+
+   If D is empty:
+     Let block_hash be the block hash of the ETH1_FOLLOW_DISTANCEth ancestor of the head of the
+       canonical eth1.0 chain.
+     Let deposit_root be the deposit root of the eth1.0 deposit contract at the block defined by
+       block_hash.
+
+   If D is nonempty:
+     Let best_vote be the member of D that has the highest vote.eth1_data.vote_count,
+       breaking ties by favoring block hashes with higher associated block height.
+     Let block_hash = best_vote.eth1_data.block_hash.
+     Let deposit_root = best_vote.eth1_data.deposit_root.
+
+   Set block.eth1_data = Eth1Data(deposit_root=deposit_root, block_hash=block_hash).
+  */
+  private Eth1Data getEth1Data(BeaconState state) {
+    List<Eth1DataVote> eth1DataVotes = state.getEth1DataVotes();
+    Eth1Data contractData =
+        depositContract.getEth1DataByDistance(chainSpec.getEth1FollowDistance());
+
+    if (eth1DataVotes.isEmpty()) {
+      return contractData;
+    }
+
+    UInt64 maxVotes =
+        eth1DataVotes.stream().map(Eth1DataVote::getVoteCount).max(UInt64::compareTo).get();
+    Eth1DataVote bestVote =
+        eth1DataVotes.stream()
+            .filter(eth1DataVote -> eth1DataVote.getVoteCount().equals(maxVotes))
+            .reduce((first, second) -> second)
+            .get();
+
+    // verify best vote data and return if verification passed,
+    // otherwise, return data from contract
+    return depositContract
+        .getEth1DataByBlockHash(bestVote.getEth1Data().getBlockHash())
+        .filter(data -> data.equals(bestVote.getEth1Data()))
+        .orElse(contractData);
+  }
+
+  private BeaconBlockBody getBlockBody(BeaconState state, PendingOperations operations) {
+    List<ProposerSlashing> proposerSlashings =
+        operations.peekProposerSlashings(chainSpec.getMaxProposerSlashings());
+    List<CasperSlashing> casperSlashings =
+        operations.peekCasperSlashings(chainSpec.getMaxCasperSlashings());
+    List<Attestation> attestations =
+        operations.peekAggregatedAttestations(chainSpec.getMaxAttestations());
+    List<Exit> exits = operations.peekExits(chainSpec.getMaxExits());
+    List<Deposit> deposits =
+        depositContract.peekDeposits(
+            chainSpec.getMaxDeposits(), state.getLatestEth1Data(), UInt64.ZERO);
```

Comment:
Don't we need to specify the latest added `Deposit` here to peek starting from it ?
Did I get it right that `state.getLatestEth1Data()` is just the 'upper bound' for searching for new deposits?

Reply:
Latest included deposit index should be a part of `BeaconState` which is not yet updated in the spec.
> Did I get it right that `state.getLatestEth1Data()` is just the 'upper bound' for searching for new deposits?

yes, exactly

Reply:
Hm... I'm hoping they include this index as `Eth1Data` since searching just by index would require maintaining additional index

Reply:
I see. Need a suggestion on keeping previous `Eth1Data` then.